### PR TITLE
feat(webpack): allow function entries for `build.transpile`

### PR DESCRIPTION
**Pull request by atinux**
## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

It's mostly a feature for module, thinking of `@nuxt/http` for example that need to add `ky` in `build.transpile` but only for legacy mode.

### Before:

```js
{
  build: {
    transpile: ['ky']
  }
}
```

### After:

```js
{
  build: {
    transpile: [
      ({ isModern }) => isModern ? null : 'ky'
    ]
  }
}
```

The main issue I see is that module author will have to check Nuxt version to be able to push a function in `build.transpile` otherwise it justs breaks 😢 

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly. (PR: https://github.com/nuxt/docs/pull/1472)
- [x] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests are passing.

### CODE REVIEW

**Review by pi0 — DISMISSED** [review]

**Comment by pi0** [comment]
> The main issue I see is that module author will have to check Nuxt version to be able to push a function in build.transpile otherwise it justs breaks cry

This is a feat/fix by next release. Nothing will be breaking.

---

Some other enhancements we could do is adding an option like `transpileLegacy` or allow object form like `{ name: 'ky', legacy: true }` or `{ name: 'ky', modern: false }`

**Review by galvez — DISMISSED** [review]

**Comment by Atinux** [comment]
I also thought of keeping a string directly and do something like `legacy:ky`, WDYT?

File: packages/webpack/src/config/base.js [hunk 1]
```diff
@@ -57,10 +58,13 @@ export default class WebpackBaseConfig {
...
     const items = [/\.vue\.js/i]
-    for (const pattern of this.buildContext.buildOptions.transpile) {
+    for (let pattern of this.buildContext.buildOptions.transpile) {
+      if (typeof pattern === 'function') {
+        pattern = pattern(this.nuxtEnv)
+      }
       if (pattern instanceof RegExp) {
         items.push(pattern)
-      } else {
+      } else if (pattern) {
```
Comment (clarkdo) [inline packages/webpack/src/config/base.js]: Should this be string ?

**Review by clarkdo — COMMENTED** [review]
If function is one of array items, then it can only return string or regex pattern.

Do you think is there any use case for multiple functions, maybe modules ?

How about just making `transpile` as function, like:

```js
transpile: () => {
  return [
    'vue-test',
    '@vue/test'
  ]
}
```

File: packages/webpack/src/config/base.js (hunk 1 above)
Comment (Atinux) [reply packages/webpack/src/config/base.js]: Yes, Will update to check `typeof pattern === 'string'`

**Comment by Atinux** [comment]
@clarkdo Making `build.transpile` a function will be harder for modules author to extend it I guess, this is why I thought about handling a push as a method (see https://github.com/nuxt/http/commit/2ce08de2ecceda27d362982eebdf0f15470f7fee#diff-d1234a869b3d648ebfcdce5a76747d71R100)

**Comment by Atinux** [comment]
Ready to review

File: packages/webpack/src/config/server.js [hunk 2]
```diff
@@ -20,10 +20,13 @@ export default class WebpackServerConfig extends WebpackBaseConfig {
     const whitelist = [
       /\.(?!js(x|on)?$)/i
     ]
-    for (const pattern of this.buildContext.buildOptions.transpile) {
+    for (let pattern of this.buildContext.buildOptions.transpile) {
```
Comment (manniL) [inline packages/webpack/src/config/server.js]: Looks like we have a duplicate code block here (as in `src/config/base.js`, L61-70). Could we extract it somehow?

File: packages/webpack/src/config/base.js [hunk 3]
```diff
@@ -57,10 +58,13 @@ export default class WebpackBaseConfig {
...
     const items = [/\.vue\.js/i]
-    for (const pattern of this.buildContext.buildOptions.transpile) {
+    for (let pattern of this.buildContext.buildOptions.transpile) {
+      if (typeof pattern === 'function') {
+        pattern = pattern(this.nuxtEnv)
+      }
       if (pattern instanceof RegExp) {
         items.push(pattern)
-      } else {
+      } else if (typeof pattern === 'string') {
```
Comment (clarkdo) [inline packages/webpack/src/config/base.js]: How about giving a warning message if pattern is not string or regexp ?

File: packages/webpack/src/config/server.js (hunk 2 above)
Comment (clarkdo) [reply packages/webpack/src/config/server.js]: Could be, but please note the difference in `path.normalize`

**Review by clarkdo — APPROVED** [review]

File: packages/webpack/src/config/server.js (hunk 2 above)
Comment (clarkdo) [reply packages/webpack/src/config/server.js]: I will do it in another pr since this has been a long time

**Pull request merged**